File: Week 22/cpp/1.ShortestPaths.cpp
/*
 * WEEK 22 - C++ ADVANCED DSA
 * Topic: Shortest Path Algorithms (Dijkstra, Bellman-Ford, Floyd-Warshall)
 * File: 1.ShortestPaths.cpp
 *
 * CONCEPT:
 *   Dijkstra:       single-source, non-negative weights, O((V+E) log V).
 *   Bellman-Ford:   single-source, allows negatives, detects neg-cycles, O(V*E).
 *   Floyd-Warshall: all-pairs DP, O(V^3).
 *
 * KEY POINTS:
 *   - Use std::priority_queue with std::greater for a min-heap.
 *   - Pair<int,int> = (dist, vertex) for the heap.
 *   - INF = INT_MAX/2 prevents overflow when summing weights.
 *
 * ALGORITHM / APPROACH:
 *   Dijkstra: lazy deletion (skip stale heap entries via dist[u] check).
 *   Bellman-Ford: V-1 relaxations + 1 negative-cycle check pass.
 *   Floyd-Warshall: triple nested loop with k as outermost.
 *
 * C++-SPECIFIC NOTES vs JAVA:
 *   - std::priority_queue is MAX-heap by default; use
 *     `priority_queue<T, vector<T>, greater<T>>` for min-heap.
 *   - Adjacency list: `vector<vector<pair<int,int>>>`.
 *   - `auto [d,u] = pq.top();` (C++17 structured bindings) for clarity.
 *
 * DRY RUN (Dijkstra, undirected V=5):
 *   Edges (0,1,4),(0,2,1),(2,1,2),(1,3,1),(2,3,5),(3,4,3).
 *   Settled order: 0(0), 2(1), 1(3), 3(4), 4(7).
 *   Final distances: [0, 3, 1, 4, 7].
 *
 * COMPLEXITY:
 *   Dijkstra:       O((V+E) log V).
 *   Bellman-Ford:   O(V*E).
 *   Floyd-Warshall: O(V^3) time, O(V^2) space.
 */

#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <climits>
#include <iomanip>

constexpr int INF = INT_MAX / 2;

std::vector<int> dijkstra(const std::vector<std::vector<std::pair<int,int>>>& adj, int src, int V) {
    std::vector<int> dist(V, INF);
    dist[src] = 0;
    using P = std::pair<int,int>; // (dist, vertex)
    std::priority_queue<P, std::vector<P>, std::greater<P>> pq;
    pq.push({0, src});
    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;
        for (auto [v, w] : adj[u]) {
            if (d + w < dist[v]) {
                dist[v] = d + w;
                pq.push({dist[v], v});
            }
        }
    }
    return dist;
}

bool bellman_ford(int V, const std::vector<std::tuple<int,int,int>>& edges, int src, std::vector<int>& dist) {
    dist.assign(V, INF);
    dist[src] = 0;
    for (int i = 0; i < V - 1; ++i) {
        bool updated = false;
        for (auto [u, v, w] : edges) {
            if (dist[u] != INF && dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                updated = true;
            }
        }
        if (!updated) break;
    }
    for (auto [u, v, w] : edges) {
        if (dist[u] != INF && dist[u] + w < dist[v]) {
            std::cout << "Negative weight cycle detected!\n";
            return false;
        }
    }
    return true;
}

std::vector<std::vector<int>> floyd_warshall(int V, const std::vector<std::tuple<int,int,int>>& edges) {
    std::vector<std::vector<int>> d(V, std::vector<int>(V, INF));
    for (int i = 0; i < V; ++i) d[i][i] = 0;
    for (auto [u, v, w] : edges) d[u][v] = std::min(d[u][v], w);
    for (int k = 0; k < V; ++k)
        for (int i = 0; i < V; ++i)
            if (d[i][k] != INF)
                for (int j = 0; j < V; ++j)
                    if (d[k][j] != INF && d[i][k] + d[k][j] < d[i][j])
                        d[i][j] = d[i][k] + d[k][j];
    for (int i = 0; i < V; ++i) if (d[i][i] < 0) { std::cout << "Negative cycle!\n"; return {}; }
    return d;
}

int main() {
    std::cout << "=== Dijkstra ===\n";
    int V = 5;
    std::vector<std::vector<std::pair<int,int>>> adj(V);
    std::vector<std::tuple<int,int,int>> edges = {
        {0,1,4},{0,2,1},{2,1,2},{1,3,1},{2,3,5},{3,4,3}
    };
    for (auto [u, v, w] : edges) {
        adj[u].push_back({v, w});
        adj[v].push_back({u, w}); // undirected
    }
    auto d = dijkstra(adj, 0, V);
    std::cout << "Shortest distances from 0: [";
    for (int i = 0; i < V; ++i) std::cout << d[i] << (i + 1 < V ? "," : "");
    std::cout << "]\n";

    std::cout << "\n=== Bellman-Ford ===\n";
    std::vector<std::tuple<int,int,int>> bf = {
        {0,1,-1},{0,2,4},{1,2,3},{1,3,2},{1,4,2},{3,2,5},{3,1,1},{4,3,-3}
    };
    std::vector<int> bfd;
    bellman_ford(5, bf, 0, bfd);
    std::cout << "Shortest distances from 0: [";
    for (size_t i = 0; i < bfd.size(); ++i) std::cout << bfd[i] << (i + 1 < bfd.size() ? "," : "");
    std::cout << "]\n";

    std::cout << "\n=== Floyd-Warshall ===\n";
    std::vector<std::tuple<int,int,int>> fw = {
        {0,1,3},{0,3,7},{1,0,8},{1,2,2},{2,0,5},{2,3,1},{3,0,2}
    };
    auto ap = floyd_warshall(4, fw);
    std::cout << "All-pairs shortest paths:\n";
    for (auto& row : ap) {
        for (int v : row) {
            if (v == INF) std::cout << std::setw(4) << "INF";
            else std::cout << std::setw(4) << v;
        }
        std::cout << "\n";
    }
    return 0;
}

File: Week 22/python/1.ShortestPaths.py
"""
WEEK 22 - PYTHON ADVANCED DSA
Topic: Shortest Path Algorithms (Dijkstra, Bellman-Ford, Floyd-Warshall)
File: 1.ShortestPaths.py

CONCEPT:
    Single-source and all-pairs shortest path algorithms on weighted graphs.
    Dijkstra:       single-source, NON-NEGATIVE weights only.
    Bellman-Ford:   single-source, allows negative weights, detects
                    negative-weight cycles.
    Floyd-Warshall: all-pairs, dynamic programming, O(V^3).

KEY POINTS:
    - Dijkstra uses a min-heap keyed by tentative distance.
    - Bellman-Ford relaxes all edges V-1 times. A V-th relaxation that
      still improves a distance proves a negative cycle is reachable.
    - Floyd-Warshall iterates over allowed intermediate vertex set
      {0..k}; recurrence d[i][j] = min(d[i][j], d[i][k] + d[k][j]).
    - Use a "large but safe" sentinel (INF // 2) so additions don't overflow.

ALGORITHM / APPROACH:
    Dijkstra(adj, src):
        dist = [INF] * V; dist[src] = 0
        push (0, src) onto heap
        while heap:
            d, u = heappop
            if d > dist[u]: continue
            for (v, w) in adj[u]:
                if d + w < dist[v]:
                    dist[v] = d + w; push (dist[v], v)

    BellmanFord(V, edges, src):
        for V-1 times: relax every edge.
        for every edge: if still relaxable -> negative cycle.

    FloydWarshall(V, edges):
        d[i][i] = 0; d[u][v] = w for each edge.
        for k in V: for i in V: for j in V:
            d[i][j] = min(d[i][j], d[i][k] + d[k][j])

PYTHON-SPECIFIC NOTES vs JAVA:
    - heapq is a min-heap; push tuples (dist, vertex).
    - math.inf is a clean sentinel; comparisons just work.
    - Adjacency list as list[list[tuple[int,int]]] reads cleanly.
    - List comprehensions replace explicit fills.

DRY RUN (Dijkstra example, undirected):
    V=5, edges {(0,1,4),(0,2,1),(2,1,2),(1,3,1),(2,3,5),(3,4,3)}
    Start dist=[0,inf,inf,inf,inf]; pq=[(0,0)].
    Pop (0,0): relax 1->4, 2->1; pq=[(1,2),(4,1)].
    Pop (1,2): relax 1->1+2=3, 3->1+5=6; pq=[(3,1),(4,1),(6,3)].
    Pop (3,1): relax 3->3+1=4; pq=[(4,1),(4,3),(6,3)].
    Pop (4,1) -> stale (3 already known); skip.
    Pop (4,3): relax 4->4+3=7; pq=[(6,3),(7,4)].
    Pop (6,3) stale; pop (7,4) -> nothing new.
    Final: [0, 3, 1, 4, 7].

COMPLEXITY:
    Dijkstra:       O((V + E) log V)
    Bellman-Ford:   O(V * E)
    Floyd-Warshall: O(V^3) time, O(V^2) space.
"""

from __future__ import annotations
import heapq
import math
from typing import List, Tuple, Optional

INF = math.inf


def dijkstra(adj: List[List[Tuple[int, int]]], src: int, V: int) -> List[float]:
    dist: List[float] = [INF] * V
    dist[src] = 0
    pq: List[Tuple[float, int]] = [(0, src)]
    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        for v, w in adj[u]:
            nd = d + w
            if nd < dist[v]:
                dist[v] = nd
                heapq.heappush(pq, (nd, v))
    return dist


def bellman_ford(V: int, edges: List[Tuple[int, int, int]], src: int) -> Optional[List[float]]:
    dist: List[float] = [INF] * V
    dist[src] = 0
    for _ in range(V - 1):
        updated = False
        for u, v, w in edges:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                updated = True
        if not updated:
            break
    # Negative cycle check
    for u, v, w in edges:
        if dist[u] + w < dist[v]:
            print("Negative weight cycle detected!")
            return None
    return dist


def floyd_warshall(V: int, edges: List[Tuple[int, int, int]]) -> Optional[List[List[float]]]:
    dist: List[List[float]] = [[INF] * V for _ in range(V)]
    for i in range(V):
        dist[i][i] = 0
    for u, v, w in edges:
        if w < dist[u][v]:
            dist[u][v] = w
    for k in range(V):
        for i in range(V):
            dik = dist[i][k]
            if dik == INF:
                continue
            row_i = dist[i]
            row_k = dist[k]
            for j in range(V):
                if dik + row_k[j] < row_i[j]:
                    row_i[j] = dik + row_k[j]
    for i in range(V):
        if dist[i][i] < 0:
            print("Negative cycle!")
            return None
    return dist


def main() -> None:
    print("=== Dijkstra ===")
    V = 5
    adj: List[List[Tuple[int, int]]] = [[] for _ in range(V)]
    edges = [(0, 1, 4), (0, 2, 1), (2, 1, 2), (1, 3, 1), (2, 3, 5), (3, 4, 3)]
    for u, v, w in edges:
        adj[u].append((v, w))
        adj[v].append((u, w))
    print("Shortest distances from 0:", dijkstra(adj, 0, V))  # [0, 3, 1, 4, 7]

    print("\n=== Bellman-Ford ===")
    bf_edges = [(0, 1, -1), (0, 2, 4), (1, 2, 3), (1, 3, 2),
                (1, 4, 2), (3, 2, 5), (3, 1, 1), (4, 3, -3)]
    print("Shortest distances from 0:", bellman_ford(5, bf_edges, 0))  # [0, -1, 2, -2, 1]

    print("\n=== Floyd-Warshall ===")
    fw_edges = [(0, 1, 3), (0, 3, 7), (1, 0, 8), (1, 2, 2),
                (2, 0, 5), (2, 3, 1), (3, 0, 2)]
    ap = floyd_warshall(4, fw_edges)
    print("All-pairs shortest paths:")
    for row in ap or []:
        print("  " + " ".join(f"{('INF' if v == INF else int(v))!s:>4}" for v in row))


if __name__ == "__main__":
    main()
